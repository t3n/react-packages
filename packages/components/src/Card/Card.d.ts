import React, { ReactNode } from 'react';
import { SizeProps, MarginProps } from 'styled-system';
export interface CardProps extends MarginProps, React.HTMLAttributes<HTMLAnchorElement | HTMLDivElement> {
    rounded?: boolean;
    big?: boolean;
    stretch?: boolean;
    elevate?: boolean;
    dashed?: boolean;
    splitted?: boolean;
    href?: string | false;
    targetBlank?: boolean;
    color?: string;
    width?: SizeProps['size'];
    children?: ReactNode;
}
export declare const Card: React.ForwardRefExoticComponent<CardProps & React.RefAttributes<HTMLDivElement | HTMLAnchorElement>>;
