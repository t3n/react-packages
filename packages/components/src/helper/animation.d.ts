import { ThemeProps } from '@t3n/theme';
export declare const backgroundAnimation: import("styled-components").FlattenInterpolation<import("styled-components").ThemedStyledProps<ThemeProps, any>>;
