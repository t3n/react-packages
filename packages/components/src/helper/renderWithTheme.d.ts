import 'jest-styled-components';
import React from 'react';
import { RenderOptions } from '@testing-library/react';
import { Theme } from '@t3n/theme';
interface OptionalTheme {
    theme?: Theme;
}
export declare const renderWithTheme: (ui: React.ReactElement, { theme, ...options }: RenderOptions & OptionalTheme) => import("@testing-library/react").RenderResult<typeof import("@testing-library/dom/types/queries"), HTMLElement>;
export {};
