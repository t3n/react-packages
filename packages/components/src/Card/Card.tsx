import { ReactNode } from 'react';
import styled from 'styled-components';
import {
  space,
  width,
  color as styledColor,
  boxShadow as styledBoxShadow,
  SizeProps,
  MarginProps
} from 'styled-system';

import { ThemeProps } from '@t3n/theme';
import { CardHeader, CardHeaderContent } from '../CardHeader';

export interface CardProps extends MarginProps {
  rounded?: boolean;
  big?: boolean;
  stretch?: boolean;
  elevate?: boolean;
  dashed?: boolean;
  splitted?: boolean;
  href?: string | false;
  color?: string;
  width?: SizeProps['size'];
  children?: ReactNode;
}

const borderRadius = ({ rounded, theme }: CardProps & ThemeProps) =>
  `border-radius: ${rounded ? theme.border.radii[1] : 0};`;

const padding = ({ big, splitted, theme }: CardProps & ThemeProps) =>
  big
    ? space({ p: [3, 6], theme })
    : splitted
    ? space({ p: 0, theme })
    : space({ p: [2, 3], theme });

const color = ({ color: c, theme }: CardProps & ThemeProps) =>
  styledColor({ color: c, theme });

const shadow = {
  default: ({ elevate, href, theme }: CardProps & ThemeProps) =>
    elevate || href ? styledBoxShadow({ boxShadow: 'elevate', theme }) : '',
  hover: ({ href, theme }: CardProps & ThemeProps) =>
    href ? styledBoxShadow({ boxShadow: 'elevateHover', theme }) : ''
};

const headerMargin = ({ big, theme }: CardProps & ThemeProps) =>
  big
    ? space({ mx: [-3, -6], mt: [-3, -6], mb: [3, 6], theme })
    : space({ mx: -3, mt: -3, mb: 3, theme });

const border = ({ dashed, elevate, href, theme }: CardProps & ThemeProps) => {
  const borderWidth = dashed && !elevate && !href ? '2px' : '1px';
  const style = dashed && !elevate && !href ? 'dashed' : 'solid';

  return `border: ${borderWidth} ${style} ${theme.colors.shades.grey204}`;
};

export const Card = styled.div.attrs(({ href }: CardProps) => ({
  href,
  as: href ? 'a' : 'div'
}))<CardProps>`
  display: block;
  background-color: white;
  display: flex;
  flex-direction: ${({ splitted }) => (splitted ? 'row' : 'column')};
  flex-wrap: ${({ splitted }) => (splitted ? 'wrap' : 'nowrap')};
  height: ${({ stretch }) => (stretch ? '100%' : 'auto')};
  justify-content: flex-start;
  justify-content: stretch;
  text-decoration: none;
  overflow: hidden;
  transition: box-shadow .15s ease-out, transform .15s ease-out;
  ${border};
  ${borderRadius}
  ${width}
  ${padding}
  ${color}
  ${shadow.default}
  ${space}

  &:hover {
    ${shadow.hover}
    ${({ href }: CardProps) =>
      href ? `transform: translate3d(0,-2px, 0);` : ''}
  }

  ${CardHeader} {
    ${headerMargin}

    ${CardHeaderContent} {
      ${padding}
    }
  }
`;

Card.displayName = 'Card';

Card.defaultProps = {
  rounded: true,
  color: 'text.primary',
  width: 1,
  mb: 2
};
