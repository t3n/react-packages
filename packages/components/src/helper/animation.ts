import { css, keyframes } from 'styled-components';

import { ThemeProps } from '@t3n/theme';

const backgroundAnimationKeyframes = ({ theme }: ThemeProps) => keyframes`
  0% {
    background-color: ${theme.colors.shades.grey204}
  }

  50% {
    background-color: ${theme.colors.shades.grey244}
  }

  100% {
    background-color: ${theme.colors.shades.grey204}
  }
`;

// eslint-disable-next-line import/prefer-default-export
export const backgroundAnimation = css`
  animation: ${backgroundAnimationKeyframes} 2s linear infinite;
`;
