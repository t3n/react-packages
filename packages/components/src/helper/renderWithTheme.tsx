/* eslint-disable import/no-extraneous-dependencies */
import 'jest-styled-components';

import React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { Theme, theme as originalTheme } from '@t3n/theme';

interface OptionalTheme {
  theme?: Theme;
}

// eslint-disable-next-line import/prefer-default-export
export const renderWithTheme = (
  ui: React.ReactElement,
  { theme, ...options }: RenderOptions & OptionalTheme
) => {
  // merge over default theme

  const mergedTheme = { ...originalTheme, ...theme };

  const Wrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={mergedTheme}>{children}</ThemeProvider>
  );

  return render(ui, { wrapper: Wrapper, ...options });
};
