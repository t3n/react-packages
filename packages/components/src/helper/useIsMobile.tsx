import { Theme } from '@t3n/theme';
import { useCallback, useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';

const useIsMobile = (): boolean => {
  const [isMobile, setIsMobile] = useState(false);

  const theme: Theme = useContext(ThemeContext);
  const fontSize = window
    .getComputedStyle(window.document.body)
    .getPropertyValue('font-size');
  const mobileBreakpoint =
    parseInt(theme.breakpoints[1], 10) * parseInt(fontSize, 10);

  const testIsMobile = useCallback(() => {
    if (window.matchMedia(`(max-width: ${mobileBreakpoint}px)`).matches) {
      setIsMobile(true);
    } else {
      setIsMobile(false);
    }
  }, [mobileBreakpoint]);

  useEffect(() => {
    testIsMobile();
    window.addEventListener('resize', testIsMobile);

    return () => {
      window.removeEventListener('resize', testIsMobile);
    };
  }, [testIsMobile]);

  return isMobile;
};

export default useIsMobile;
