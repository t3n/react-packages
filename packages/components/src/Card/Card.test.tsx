import React from 'react';

import { renderWithTheme } from '../helper/renderWithTheme';
import { Card } from './Card';

test('Default Card matches snapshot', () => {
  const { container } = renderWithTheme(<Card>Default Card</Card>, {});

  expect(container.firstChild).toMatchSnapshot();
});

test('Big Card matches snapshot', () => {
  const { container } = renderWithTheme(<Card big>Big card</Card>, {});
  expect(container.firstChild).toMatchSnapshot();
});

test('Dashed Card matches snapshot', () => {
  const { container } = renderWithTheme(<Card dashed>Dashed card</Card>, {});
  expect(container.firstChild).toMatchSnapshot();
});
