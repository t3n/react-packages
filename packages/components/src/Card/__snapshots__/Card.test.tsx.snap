// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Big Card matches snapshot 1`] = `
.c0 {
  display: block;
  background-color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  height: auto;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  -webkit-text-decoration: none;
  text-decoration: none;
  overflow: hidden;
  -webkit-transition: box-shadow .15s ease-out,-webkit-transform .15s ease-out;
  -webkit-transition: box-shadow .15s ease-out,transform .15s ease-out;
  transition: box-shadow .15s ease-out,transform .15s ease-out;
  border: 1px solid #cccccc;
  border-radius: 4px;
  width: 100%;
  padding: 24px;
  color: #2a2a2a;
  margin-bottom: 16px;
}

.c0 .c2 {
  margin-left: -24px;
  margin-right: -24px;
  margin-top: -24px;
  margin-bottom: 24px;
}

.c0 .c2 .c1 {
  padding: 24px;
}

@media screen and (min-width:30rem) {
  .c0 {
    padding: 48px;
  }
}

@media screen and (min-width:30rem) {
  .c0 .c2 {
    margin-left: -48px;
    margin-right: -48px;
    margin-top: -48px;
    margin-bottom: 48px;
  }
}

@media screen and (min-width:30rem) {
  .c0 .c2 .c1 {
    padding: 48px;
  }
}

<div
  class="c0"
  color="text.primary"
  width="1"
>
  Big card
</div>
`;

exports[`Dashed Card matches snapshot 1`] = `
.c0 {
  display: block;
  background-color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  height: auto;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  -webkit-text-decoration: none;
  text-decoration: none;
  overflow: hidden;
  -webkit-transition: box-shadow .15s ease-out,-webkit-transform .15s ease-out;
  -webkit-transition: box-shadow .15s ease-out,transform .15s ease-out;
  transition: box-shadow .15s ease-out,transform .15s ease-out;
  border: 2px dashed #cccccc;
  border-radius: 4px;
  width: 100%;
  padding: 16px;
  color: #2a2a2a;
  margin-bottom: 16px;
}

.c0 .c2 {
  margin-left: -24px;
  margin-right: -24px;
  margin-top: -24px;
  margin-bottom: 24px;
}

.c0 .c2 .c1 {
  padding: 16px;
}

@media screen and (min-width:30rem) {
  .c0 {
    padding: 24px;
  }
}

@media screen and (min-width:30rem) {
  .c0 .c2 .c1 {
    padding: 24px;
  }
}

<div
  class="c0"
  color="text.primary"
  width="1"
>
  Dashed card
</div>
`;

exports[`Default Card matches snapshot 1`] = `
.c0 {
  display: block;
  background-color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  height: auto;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  -webkit-text-decoration: none;
  text-decoration: none;
  overflow: hidden;
  -webkit-transition: box-shadow .15s ease-out,-webkit-transform .15s ease-out;
  -webkit-transition: box-shadow .15s ease-out,transform .15s ease-out;
  transition: box-shadow .15s ease-out,transform .15s ease-out;
  border: 1px solid #cccccc;
  border-radius: 4px;
  width: 100%;
  padding: 16px;
  color: #2a2a2a;
  margin-bottom: 16px;
}

.c0 .c2 {
  margin-left: -24px;
  margin-right: -24px;
  margin-top: -24px;
  margin-bottom: 24px;
}

.c0 .c2 .c1 {
  padding: 16px;
}

@media screen and (min-width:30rem) {
  .c0 {
    padding: 24px;
  }
}

@media screen and (min-width:30rem) {
  .c0 .c2 .c1 {
    padding: 24px;
  }
}

<div
  class="c0"
  color="text.primary"
  width="1"
>
  Default Card
</div>
`;
